File: 1961-maximum-ice-cream-bars/maximum-ice-cream-bars.py
class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:
        # 1. finding the maximum
        max_val = max(costs)
        # 2. create frequency list
        cnt = [0] * (max_val + 1)
        # 3. count the frequency
        for num in costs: 
            cnt[num] += 1
        # 4. try to get cheapest item from the frequency info
        ans = 0
        for price in range(1, max_val+1):
            
            # can not buy anymore (the more iteration the more expansive)
            if price > coins:
                break

            if cnt[price] == 0:
                continue
            
            # how many can buy at this price
            # when the supplies can only offer small amount, choose this
            can_buy = min(cnt[price], coins // price)

            coins -= can_buy * price
            ans += can_buy
        
        return ans

File: 1961-maximum-ice-cream-bars/maximum-ice-cream-bars.cpp
class Solution {
public:
    int maxIceCream(vector<int>& costs, int coins) {
        int n = costs.size();
        // 1. finding the maximum
        int max_val = 0;
        for (int i=0; i<n; i++) max_val = max(max_val, costs[i]);
        // 2. create a array of size max+1;
        vector<int> cnt(max_val+1, 0);
        // 3. traverse the original list and counting the frequency
        for (int i=0; i<n; i++) cnt[costs[i]]++;
        // 4. calculate the prefix sum
        for (int i=1; i<max_val+1; i++) cnt[i] = cnt[i-1] + cnt[i];
        // 5. sorting the array
        vector<int> res(n, 0);
        for (int i=n-1; i>=0; i--) {
            res[cnt[costs[i]] - 1] = costs[i];
            cnt[costs[i]]--;
        }
        // 6. traverse the answer and find the maximum number of ice cream
        int ans = 0;
        for (int i=0; i<n; i++) {
            if (coins >= res[i]) {
                coins -= res[i];
                ans++;
            }
        }
        return ans;
    }
};